synapse/handlers/federation.py

```diff
@@ -1035,6 +1043,217 @@ async def get_state_ids_for_pdu(self, room_id: str, event_id: str) -> List[str]:
+                        seed_event.type,
+                    )
+                )
+
+        while not queue.empty() and len(event_id_results) < limit:
+            try:
+                _, _, event_id, event_type = queue.get_nowait()
+            except Empty:
+                break
+
+            if event_id in event_id_results:
+                continue
+
+            found_undiscovered_connected_historical_messages = False
+            if self.hs.config.experimental.msc2716_enabled:
+                # Try and find any potential historical batches of message history.
+                #
+                # First we look for an insertion event connected to the current
+                # event (by prev_event). If we find any, we'll add them to the queue
+                # and navigate up the DAG like normal in the next iteration of the
+                # loop.
+                connected_insertion_event_backfill_results = (
+                    await self.store.get_connected_insertion_event_backfill_results(
+                        event_id, limit - len(event_id_results)
+                    )
+                )
+                logger.info(
+                    "get_backfill_events(room_id=%s): connected_insertion_event_backfill_results(%s)=%s",
+                    room_id,
+                    event_id,
+                    connected_insertion_event_backfill_results,
+                )
+                for (
+                    connected_insertion_event_backfill_item
+                ) in connected_insertion_event_backfill_results:
+                    if (
+                        connected_insertion_event_backfill_item.event_id
+                        not in event_id_results
+                    ):
+                        # Check whether the insertion event is already on the
+                        # federating homeserver we're trying to send backfill
+                        # events to
+                        room_version = await self.store.get_room_version(room_id)
+                        event_exists_on_remote_server = None
+                        try:
+                            # Because of the nature of backfill giving events to
+                            # the federated homeserver in one chunk and then we
+                            # can possibly query about that same event in the
+                            # next chunk, we need to avoid getting a cached
+                            # response. We want to know *now* whether they have
+                            # backfilled the insertion event.
+                            event_exists_on_remote_server = await self.federation_client.get_pdu_from_destination_raw(
+                                origin,
+                                connected_insertion_event_backfill_item.event_id,
+                                room_version=room_version,
+                                outlier=True,
+                                timeout=10000,
+                            )
```

Comment:
Refactored `get_backfill_events` out to the handler to address https://github.com/matrix-org/synapse/pull/11114#discussion_r731363046

We need to be able to ask the federating homeserver whether they have the insertion event we're potentially going to fork on.

 - If the federating homserver does not have the insertion event, we go down the historical branch
 - If the federating homserver already has the insertion event, we ignore the historical branch and go down the normal DAG below